internal/controllers/machineset/machineset_controller_status.go

```diff
@@ -0,0 +1,327 @@
+	// If we got unexpected errors in listing the machines (this should never happen), surface them.
+	if !getAndAdoptMachinesForMachineSetSucceeded {
+		v1beta2conditions.Set(machineSet, metav1.Condition{
+			Type:    clusterv1.MachineSetMachinesUpToDateV1Beta2Condition,
+			Status:  metav1.ConditionUnknown,
+			Reason:  clusterv1.MachineSetMachinesUpToDateInternalErrorV1Beta2Reason,
+			Message: "Please check controller logs for errors",
+		})
+		return
+	}
+
+	if len(machines) == 0 {
+		v1beta2conditions.Set(machineSet, metav1.Condition{
+			Type:   clusterv1.MachineSetMachinesUpToDateV1Beta2Condition,
+			Status: metav1.ConditionTrue,
+			Reason: clusterv1.MachineSetMachinesUpToDateNoReplicasV1Beta2Reason,
+		})
+		return
+	}
+
+	upToDateCondition, err := v1beta2conditions.NewAggregateCondition(
+		machines, clusterv1.MachineUpToDateV1Beta2Condition,
+		v1beta2conditions.TargetConditionType(clusterv1.MachineSetMachinesUpToDateV1Beta2Condition),
+	)
+	if err != nil {
+		log := ctrl.LoggerFrom(ctx)
+		log.Error(err, "Failed to aggregate Machine's UpToDate conditions")
+		v1beta2conditions.Set(machineSet, metav1.Condition{
+			Type:    clusterv1.MachineSetMachinesUpToDateV1Beta2Condition,
+			Status:  metav1.ConditionUnknown,
+			Reason:  clusterv1.MachineSetMachinesUpToDateInternalErrorV1Beta2Reason,
+			Message: "Please check controller logs for errors",
+		})
+		return
+	}
+
+	v1beta2conditions.Set(machineSet, *upToDateCondition)
+}
+
+func calculateMissingReferencesMessage(ms *clusterv1.MachineSet, bootstrapTemplateNotFound, infraMachineTemplateNotFound bool) string {
+	missingObjects := []string{}
+	if bootstrapTemplateNotFound {
+		missingObjects = append(missingObjects, ms.Spec.Template.Spec.Bootstrap.ConfigRef.Kind)
+	}
+	if infraMachineTemplateNotFound {
+		missingObjects = append(missingObjects, ms.Spec.Template.Spec.InfrastructureRef.Kind)
+	}
+
+	if len(missingObjects) == 0 {
+		return ""
+	}
+
+	if len(missingObjects) == 1 {
+		return fmt.Sprintf("because %s does not exist", missingObjects[0])
+	}
+
+	return fmt.Sprintf("because %s do not exist", strings.Join(missingObjects, " and "))
+}
+
+func aggregateStaleMachines(machines []*clusterv1.Machine) string {
+	machineNames := []string{}
+	for _, machine := range machines {
+		if !machine.GetDeletionTimestamp().IsZero() && time.Since(machine.GetDeletionTimestamp().Time) >= time.Minute*30 {
+			machineNames = append(machineNames, machine.GetName())
+		}
+	}
+
+	if len(machineNames) == 0 {
+		return ""
+	}
+
+	message := "Machine"
+	if len(machineNames) > 1 {
+		message += "s"
+	}
+
+	sort.Strings(machineNames)
+	message += " " + clog.ListToString(machineNames, func(s string) string { return s }, 3)
+
+	if len(machineNames) == 1 {
+		message += " is "
+	} else {
+		message += " are "
+	}
+	message += "in deletion by more than 30 minutes"
```

Comment:
```suggestion
	message += "in deletion since more than 30m"
```
Just to make it slightly more consistent with the Machine controller